jax/core.py

```diff
@@ -1134,6 +1134,7 @@ def join(self, other):
     else:
       assert False, f"Cannot join {self} with {other}"
   def str_short(self): return 'Tok'
+  def at_least_vspace(self): return self
```

Comment:
It's not clear to me when this should be called (if ever), and what it should produce if it is called. But if it helps fix an issue then it seems pretty harmless.

Reply:
There's a stack trace in https://github.com/google/jax/issues/5463. The salient bits seem to be:

```
../../../../../Documents/pythonenvs/mpi4jax_env/lib/python3.8/site-packages/jax/interpreters/partial_eval.py:506: in trace_to_jaxpr
    jaxpr, (out_pvals, consts, env) = fun.call_wrapped(pvals)
../../../../../Documents/pythonenvs/mpi4jax_env/lib/python3.8/site-packages/jax/linear_util.py:160: in call_wrapped
    ans = self.f(*args, **dict(self.params, **kwargs))
tests/test_collective_ops.py:202: in <lambda>
    res, vjp_fun = jax.vjp(lambda x: Allreduce(x, op=MPI.SUM, token=token)[0], arr)
mpi4jax/validation.py:90: in wrapped
    return function(*args, **kwargs)
mpi4jax/collective_ops/allreduce.py:68: in Allreduce
    return mpi_allreduce_p.bind(x, token, op=op, comm=comm, transpose=_transpose)
../../../../../Documents/pythonenvs/mpi4jax_env/lib/python3.8/site-packages/jax/core.py:270: in bind
    tracers = map(top_trace.full_raise, args)
../../../../../Documents/pythonenvs/mpi4jax_env/lib/python3.8/site-packages/jax/_src/util.py:37: in safe_map
    return list(map(f, *args))
../../../../../Documents/pythonenvs/mpi4jax_env/lib/python3.8/site-packages/jax/core.py:377: in full_raise
    return self.pure(val)
../../../../../Documents/pythonenvs/mpi4jax_env/lib/python3.8/site-packages/jax/interpreters/ad.py:269: in pure
    tangent_zero = Zero(get_aval(val).at_least_vspace())
```

Is this still a sensible fix?

Reply:
I see! That's very informative.

And you're right, it makes sense. We have a primitive that takes a float and a token as arguments. Both could be Tracers, e.g. the former in a JVPTrace and the latter in a JaxprTrace. The way we process traces, we [raise all args into the trace in question](https://github.com/google/jax/blob/9ccfc9fd48b097c78c2d0fae515ff6bd52c0f681/jax/core.py#L270) with trivial context; in this case that means lifting the token into the JVPTrace with a trivial (zero) tangent.

So two viable solutions:
1. make the standard `bind` work by allowing tokens to be lifted into the JVPTrace with trivial context, which ultimately means making `AbstractToken.at_least_vspace` work;
2. define a custom bind for this primitive which doesn't try to lift its token argument.

The first option seems reasonable to me, and lighter weight! So yes this seems like a sensible fix.

Thanks for the context!